File: PixivRankingHandler.py
import os
import sys

from colorama import Fore, Style

import PixivConfig
import PixivConstant
import PixivException
import PixivHelper
import PixivImageHandler
from PixivBrowserFactory import PixivBrowser
from PixivListItem import PixivListItem


def process_ranking(caller, config, mode, content, start_page=1, end_page=0, date="", filter=None, notifier=None):
    if notifier is None:
        notifier = PixivHelper.dummy_notifier
    msg = Fore.YELLOW + Style.NORMAL + f'Processing Pixiv Ranking: {mode}.' + Style.RESET_ALL
    PixivHelper.print_and_log(None, msg)
    br: PixivBrowser = caller.__br__

    try:
        # get rank data
        curr_page = start_page
        i = 1
        while(True):
            title_prefix = f"Pixiv Ranking: {mode} page {curr_page} and date {date}"
            PixivHelper.print_and_log(None, title_prefix)
            ranks = br.getPixivRanking(mode, curr_page, date, content, filter)
            print(f"Mode : {ranks.mode}")
            print(f"Total: {ranks.rank_total}")
            print(f"Next Page: {ranks.next_page}")
            print(f"Next Date: {ranks.next_date}")

            ignore_file_list = "ignore_list.txt"
            if os.path.exists(ignore_file_list):
                PixivHelper.print_and_log('info', f'Using ignore list for member: {ignore_file_list}')
                ignore_list = PixivListItem.parseList(ignore_file_list, config.rootDirectory)
                for ignore in ignore_list:
                    for item in ranks.contents:
                        if item["user_id"] == ignore.memberId:
                            ranks.remove(item)
                            break

            for post in ranks.contents:
                try:
                    result = PixivConstant.PIXIVUTIL_OK
                    if not caller.DEBUG_SKIP_PROCESS_IMAGE:
                        print(f"{Fore.YELLOW} #{i}.{Style.RESET_ALL}")
                        result = PixivImageHandler.process_image(caller,
                                                                config,
                                                                None,
                                                                post["illust_id"],
                                                                user_dir=config.rootDirectory,
                                                                title_prefix=title_prefix,
                                                                image_response_count=post["rating_count"],
                                                                notifier=notifier)
                    PixivHelper.wait(result, config)
                    i = i + 1
                except PixivException as pex:
                    PixivHelper.print_and_log("error", f"Failed to process PixivRanking for {post.illust_id} ==> {pex.message}")

            curr_page = curr_page + 1
            if end_page > 0 and curr_page > end_page:
                PixivHelper.print_and_log("info", f"Reach max page {end_page}")
                break
            if type(ranks.next_page) is not int:
                PixivHelper.print_and_log("info", "Reaching last page.")
                break

    except Exception as ex:
        if isinstance(ex, KeyboardInterrupt):
            raise
        caller.ERROR_CODE = getattr(ex, 'errorCode', -1)
        PixivHelper.print_and_log('error', f'Error at process_ranking(): {sys.exc_info()}')
        print('Failed')
        raise


def process_new_illusts(caller, config: PixivConfig, type_mode="illust", max_page=0, notifier=None):
    if notifier is None:
        notifier = PixivHelper.dummy_notifier
    br: PixivBrowser = caller.__br__

    msg = Fore.YELLOW + Style.NORMAL + f'Processing Pixiv Ranking: RR-18={config.r18mode}.' + Style.RESET_ALL
    PixivHelper.print_and_log(None, msg)
    last_id = 0
    current_page = 1
    i = 1
    while True:
        result = br.getNewIllust(last_id, type_mode=type_mode, r18=config.r18mode)
        title_prefix = f"Pixiv New Illusts: RR-18={config.r18mode} - page {current_page}."
        PixivHelper.print_and_log(None, title_prefix)

        for image in result.images:
            try:
                dl_result = PixivConstant.PIXIVUTIL_OK
                image_id = image["id"]
                print(f"{Fore.YELLOW} #{i} - {image_id}.{Style.RESET_ALL}")
                if not caller.DEBUG_SKIP_PROCESS_IMAGE:
                    dl_result = PixivImageHandler.process_image(caller,
                                                            config,
                                                            None,
                                                            image_id,
                                                            user_dir=config.rootDirectory,
                                                            title_prefix=title_prefix,
                                                            notifier=notifier)
                PixivHelper.wait(dl_result, config)
                i = i + 1
            except PixivException as pex:
                PixivHelper.print_and_log("error", f"Failed to process Pixiv New Illusts for {image_id} ==> {pex.message}")

        last_id = result.last_id
        current_page = current_page + 1

        if max_page != 0 and current_page > max_page:
            PixivHelper.print_and_log("info", f"Reached max page = {max_page}.")
            break
        elif last_id == 0 or len(result.images) == 0:
            PixivHelper.print_and_log("info", "No more images!")
            break

File: lib/handler/pixiv_ranking_handler.dart
/// Ranking handler.
library;

import '../common/pixiv_browser.dart';
import '../common/pixiv_config.dart';
import '../common/pixiv_constant.dart' as pixiv_constant;
import '../common/pixiv_exception.dart';
import '../common/pixiv_helper.dart' as pixiv_helper;
import 'pixiv_image_handler.dart' as image_handler;

const int rankingOk = pixiv_constant.PIXIVUTIL_OK;

Future<void> processRanking({
  required dynamic caller,
  required PixivConfig config,
  required String mode,
  required String content,
  int startPage = 1,
  int endPage = 0,
  String date = '',
  List<String>? filter,
  void Function({String? title, String? message, dynamic type})? notifier,
}) async {
  notifier ??= pixiv_helper.dummyNotifier;
  pixiv_helper.printAndLog(
      'info', 'Processing Pixiv Ranking: $mode.');
  final br = caller.br as PixivBrowser;

  var currPage = startPage;
  while (true) {
    final titlePrefix =
        'Pixiv Ranking: $mode page $currPage and date $date';
    pixiv_helper.printAndLog(null, titlePrefix);
    final ranks =
        await br.getPixivRanking(mode, currPage, date, content, filter);
    pixiv_helper.printAndLog(null, 'Mode      : ${ranks.mode}');
    pixiv_helper.printAndLog(null, 'Total     : ${ranks.rankTotal}');
    pixiv_helper.printAndLog(null, 'Next Page : ${ranks.nextPage}');

    for (final post in ranks.contents) {
      try {
        final result = await image_handler.processImage(
          caller: caller,
          config: config,
          imageId: (post['illust_id'] as num).toInt(),
          userDir: config.rootDirectory,
          titlePrefix: titlePrefix,
          imageResponseCount: (post['rating_count'] as num? ?? 0).toInt(),
          notifier: notifier,
        );
        await pixiv_helper.wait(result, config);
      } on PixivException catch (e) {
        pixiv_helper.printAndLog(
            'error', 'Failed ranking image ${post['illust_id']}: $e');
      }
    }
    currPage++;
    if (endPage > 0 && currPage > endPage) break;
    if (ranks.nextPage is! int) break;
  }
}

Future<void> processNewIllusts({
  required dynamic caller,
  required PixivConfig config,
  String typeMode = 'illust',
  int maxPage = 0,
  void Function({String? title, String? message, dynamic type})? notifier,
}) async {
  notifier ??= pixiv_helper.dummyNotifier;
  final br = caller.br as PixivBrowser;
  var lastId = 0;
  var currentPage = 1;
  var i = 1;
  while (true) {
    final result = await br.getNewIllust(lastId,
        typeMode: typeMode, r18: config.r18mode);
    pixiv_helper.printAndLog(null,
        'Pixiv New Illusts: R-18=${config.r18mode} - page $currentPage');
    for (final image in result.images ?? const []) {
      final imageId = (image['id'] as num).toInt();
      pixiv_helper.printAndLog(null, '#$i - $imageId');
      try {
        final dl = await image_handler.processImage(
          caller: caller,
          config: config,
          imageId: imageId,
          userDir: config.rootDirectory,
          titlePrefix:
              'Pixiv New Illusts: R-18=${config.r18mode} - page $currentPage',
          notifier: notifier,
        );
        await pixiv_helper.wait(dl, config);
      } on PixivException catch (e) {
        pixiv_helper.printAndLog(
            'error', 'Failed new illusts image $imageId: $e');
      }
      i++;
    }
    lastId = result.lastId;
    currentPage++;
    if (maxPage != 0 && currentPage > maxPage) break;
    if (lastId == 0 || (result.images?.isEmpty ?? true)) break;
  }
}

